backdoor/src/main/java/org/codice/acdebugger/backdoor/Backdoor.java

```diff
@@ -0,0 +1,479 @@
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.acdebugger.backdoor;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.security.AccessController;
+import java.security.Permission;
+import java.security.PrivilegedAction;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+import java.security.ProtectionDomain;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.codice.acdebugger.PermissionService;
+import org.codice.acdebugger.common.JsonUtils;
+import org.codice.acdebugger.common.ServicePermissionInfo;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleActivator;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.BundleReference;
+import org.osgi.framework.Constants;
+import org.osgi.framework.ServiceEvent;
+import org.osgi.framework.ServicePermission;
+import org.osgi.framework.ServiceReference;
+import org.osgi.framework.wiring.BundleWiring;
+import org.osgi.util.tracker.ServiceTracker;
+
+/** Provides a backdoor access point for the AC debugger. */
+public class Backdoor implements BundleActivator {
+  @SuppressWarnings({
+    "squid:S1068" /* DO NOT CHANGE THIS NAME, the AC debugger is accessing it directly */
+  })
+  private static volatile Backdoor instance = null;
+
+  private volatile ServiceTracker<PermissionService, PermissionService> permServiceTracker = null;
+
+  @SuppressWarnings({
+    "squid:S2696" /* singleton instance set when initialized so AC debugger can easily find it */
+  })
+  @Override
+  public void start(BundleContext bundleContext) {
+    this.permServiceTracker =
+        new ServiceTracker<>(bundleContext, PermissionService.class.getName(), null);
+    this.permServiceTracker.open();
+    Backdoor.instance = this;
+  }
+
+  @Override
+  public void stop(BundleContext bundleContext) {
+    permServiceTracker.close();
+  }
+
+  /**
+   * Gets a bundle location for the given object. The object can be a bundle, a classloader, a
+   * protection domain, a bundle context, or even a classloader. This methods makes all attempts
```

Comment:
Classloader is mentioned twice here. Was it meant to be something else or just a typo?